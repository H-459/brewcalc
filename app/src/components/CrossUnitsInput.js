import React from "react";
import PropTypes from "prop-types";
import { convert } from "brewcalc";

const unitLabelMap = {
  sg: "SG",
  plato: "°P",
  F: "°F",
  C: "°C"
};

const stringifyMeasurable = (measurable, precision) => {
  const unit = unitLabelMap[measurable.unit] || measurable.unit;
  const value =
    precision != null ? +measurable.value.toFixed(precision) : measurable.value;
  return `${value} ${unit}`;
};

const convertMeasurable = (measurable, unit, precision) => {
  return {
    value: convert(measurable.value, measurable.unit, unit, precision),
    unit: unit
  };
};

export const printMeasurable = (measurable, convertTo, precision = 0) => {
  if (convertTo == null) {
    return stringifyMeasurable(measurable, precision);
  }
  return stringifyMeasurable(
    convertMeasurable(measurable, convertTo, precision)
  );
};

const CrossUnitsInput = ({
  name,
  description,
  measurable,
  units = [],
  precision = 2
}) => {
  if (measurable == null || measurable.value == null) {
    return null;
  }

  const [primary, secondary] = units;

  return (
    <div title={description}>
      <b>{name} </b>
      {printMeasurable(measurable, primary, precision)}
      {secondary && ` (${printMeasurable(measurable, secondary, precision)})`}
    </div>
  );
};

CrossUnitsInput.propTypes = {
  name: PropTypes.string,
  description: PropTypes.string,
  measurable: PropTypes.shape({
    value: PropTypes.number,
    unit: PropTypes.string
  }),
  units: PropTypes.arrayOf(PropTypes.string),
  precision: PropTypes.arrayOf(PropTypes.number) || PropTypes.number
};

export default CrossUnitsInput;
