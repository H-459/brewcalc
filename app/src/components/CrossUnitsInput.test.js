import React from "react";
import renderer from "react-test-renderer";
import CrossUnitsInput from "./CrossUnitsInput";

test("renders without crashing", () => {
  const tree = renderer
    .create(
      <CrossUnitsInput
        description="description"
        name="Boil Size"
        measurable={{ value: 20, unit: "l" }}
        units={["l", "gal"]}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders without crashing on US units", () => {
  const tree = renderer
    .create(
      <CrossUnitsInput
        description="description"
        name="Boil Size"
        measurable={{ value: 5, unit: "gal" }}
        units={["l", "gal"]}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
