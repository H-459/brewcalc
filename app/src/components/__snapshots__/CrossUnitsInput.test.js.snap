// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  title="description"
>
  <b>
    Boil Size
     
  </b>
  20 l
   (5.28 gal)
</div>
`;

exports[`renders without crashing on US units 1`] = `
<div
  title="description"
>
  <b>
    Boil Size
     
  </b>
  18.93 l
   (5 gal)
</div>
`;
